File: problems/242.ValidAnagram/main.ts
import { assertEquals } from "@std/assert/equals";

Deno.test("problem 242", () => {
  assertEquals(isAnagram("anagram", "nagaram"), true);
  assertEquals(isAnagram("rat", "car"), false);
});

function isAnagram(s: string, t: string): boolean {
  if (s.length !== t.length) {
    return false;
  }
  const arr = new Array(26).fill(0);
  const charA = "a".charCodeAt(0);
  for (let i = 0; i < s.length; i++) {
    const idxA = s.charCodeAt(i) - charA;
    const idxB = t.charCodeAt(i) - charA;
    arr[idxA]++;
    arr[idxB]--;
  }
  return arr.every((x) => x === 0);
}

File: problems/242.ValidAnagram/main.go
package problems

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	arr := make([]int, 26)
	charA := int('a')

	for i := 0; i < len(s); i++ {
		idxA := int(s[i]) - charA
		idxB := int(t[i]) - charA
		arr[idxA]++
		arr[idxB]--
	}

	for _, count := range arr {
		if count != 0 {
			return false
		}
	}

	return true
}
